File: libraries/Combinatorics.php
<?php
/**
 * Combinatorics methods for strings and integers
 */

class Combinatorics // no public/private/protected keyword required
{
	/**
	* Inserts a character into a string
	* @param string $str, string $ch (of length 1), integer $i
	* @return string
	*/
	private static function insert_char_at($str, $ch, $i)
	{
		if ($i === 0) return $ch . $str;
		if ($i === strlen($str)) return $str. $ch;
		else return substr($str, 0, $i) . $ch . substr($str, $i);
	}
	
	/**
	* Removes a character from a string
	* @param string $str, integer $i
	* @return string
	*/
	private static function remove_char_at($str, $i)
	{
		return substr($str, 0, $i) . substr($str, $i + 1);
	}
	
	/**
	* Gets new strings formed by adding one character to an existing string
	* @param string $str, string $ch (of length 1)
	* @return array of strings
	*/
	private static function all_new_strings1($str, $ch) // PHP does not support overloaded functions
	{
		$result = array();
		for ($i = 0; $i < strlen($str) + 1; $i++)
		{
			$result[] = self::insert_char_at($str, $ch, $i);
		}
		return $result;
	}
	
	/**
	* Gets new strings formed by adding one character to an existing array of strings
	* @param array of strings $strings, string $ch (of length 1)
	* @return array of strings
	*/
	private static function all_new_strings2($strings, $ch)
	{
		$p = count($strings);
		$q = strlen($strings[0]) + 1;
		$result = array();
		$j = -1;
		for ($i = 0; $i < $p * $q; $i++)
		{
			if ($i % $q === 0) $j++;
			$result[] = self::all_new_strings1($strings[$j], $ch)[$i % $q]; 
		}
		return $result;
	}
	
	/**
	* Suppose that we enumerate all k-combinations of a set S of n elements in lexicographic order e.g. if S = {1, 2, 3, 4, 5}, k = 3, then we have {1, 2, 3}, {1, 	2, 4}, {1, 2, 5}, {1, 3, 4}, {1, 3, 5}, {1, 4, 5}, {2, 3, 4}, {2, 3, 5}, {2, 4, 5}, {3, 4, 5}. Given one of these k-combinations, returns the next k-combination 	e.g. next({1, 4, 5}, 5) returns {2, 3, 4}
	* @param array of integers $T, integer $n
	* @return array of integers
	*/
	public static function next($T, $n)
	{
		$result = array();
		for ($i = 0; $i < count($T); $i++)
		{
			$result[] = $T[$i];
		}
		$i = 0;
		while (count($T) - 1 - $i > -1)
		{
			if ($T[count($T) - 1 - $i] < $n - $i)
			{
				$result[count($result) - 1 - $i]++;
				for ($j = count($result) - $i; $j < count($T); $j++)
				{
					$result[$j] = $result[count($result) - 1 - $i] + 1 + ($j - (count($result) - $i));
				}
				break;
			}
			$i++;
		}
		return $result;
	}
	
	/**
	* Enumerates all k-combinations of a set containing n elements in lexicographic order e.g. enum_n_choose_k(3, 5) returns {1, 2, 3}, {1, 	2, 4}, {1, 2, 5}, 	{1, 3, 4}, {1, 3, 5}, {1, 4, 5}, {2, 3, 4}, {2, 3, 5}, {2, 4, 5}, {3, 4, 5}
	* @param integer $k, integer $n
	* @return two dimensional array of integers
	*/
	public static function enum_n_choose_k($k, $n) 
	{
		$result = array();
		$first_enum = array();
		for ($i = 1; $i < $k + 1; $i++)
		{
			$first_enum[] = $i;
		}
		$result[] = $first_enum;
		for ($i = 0; $i < self::n_choose_k($k, $n) - 1; $i++)
		{
			$result[] = self::next($result[$i], $n);
		}
		return $result;
	}
	
	/**
	* Calculates the factorial of a number
	* @param integer $n
	* @return integer
	*/
	public static function fact($n)
	{
		if ($n === 0) return 1;
		$result = 1;
		for ($i = 1; $i <= $n; $i++)
		{
			$result *= $i;
		}
		return $result;
	}
	
	/**
	* Calculates n choose k
	* @param integer $k, integer $n
	* @return integer
	*/
	public static function n_choose_k($k, $n)
	{
		if ($k === 0) return 1;
		$result = 1;
		for ($i = 1; $i < $k + 1; $i++) {
			$result *= $n - ($k - $i);
			$result = floor($result / $i);
		}
		return $result;
	}
	
	/**
	* Checks if two numbers are permutations of one another
	* @param integer $a, integer $b
	* @return boolean
	*/
	public static function is_perm($a, $b)
	{
		$aAsStr = strval($a);
		$bAsStr = strval($b);
		if (strlen($aAsStr) !== strlen($bAsStr)) return false;
		$A = array();
		$B = array();
		for ($i = 0; $i < strlen($aAsStr); $i++)
		{
			$A[intval(substr($aAsStr, $i, 1))]++;
			$B[intval(substr($bAsStr, $i, 1))]++;
		}
		for ($i = 0; $i < 10; $i++)
		{
			if ($A[$i] !== $B[$i]) return false;
		}
		return true;
	}
	
	/**
	* Gets all permutations of a string
	* @param string $str
	* @return array of strings
	*/
	public static function permutations_of_a_string($str)
	{
		if (strlen($str) == 2) 
		{
			$result = array();
			$result[0] = $str;
			$result[1] = substr($str, 1) . substr($str, 0, 1);
			return $result;
		}
		return self::all_new_strings2(
										self::permutations_of_a_string(self::remove_char_at($str, strlen($str) - 1)), 
									 	substr($str, strlen($str) - 1)
									);
	}	
}
?>

File: libraries/combinatorics.cpp
#include <iostream>
#include <string>
#include "combinatorics.h"
/**
 * Please refer to combinatorics.h for function descriptions
 */

// Auxiliary functions

std::string insertCharAt(std::string str, std::string ch, int i) {
	if (i == 0) return ch + str;
	if (i == str.length()) return str + ch;
	else return str.substr(0, i) + ch + str.substr(i);	
}

std::string removeCharAt(std::string str, int i) {
	return str.substr(0, i) + str.substr(i + 1);
}

std::string * allNewStrings(std::string str, std::string ch) {
	std::string *result = new std::string[str.length() + 1];
	for (int i = 0; i < str.length() + 1; i++) {
		result[i] = insertCharAt(str, ch, i);
	}
	return result;
}

std::string * allNewStrings(std::string T[], std::string ch, int l) { // l is the length of the array T
	int k = T[0].length() + 1;
	std::string *result = new std::string[l * k]; 
	int j = -1;
	for (int i = 0; i < l * k; i++) {
		if (i % k == 0) j++;
		result[i] = allNewStrings(T[j], ch)[i % k];			
	}
	return result;
}

// Functions

int * next(int T[], int t, int n) { 
	int * result = new int[t];
	for (int i = 0; i < t; i++) {
		result[i] = T[i];
	}
	int i = 0;
	while (t - 1 - i > -1) {
		if (T[t - 1 - i] < n - i) {
			result[t - 1 - i]++;
			for (int j = t - i; j < t; j++) {
				result[j] = result[t - 1 - i] + 1 + (j - (t - i));
			}
			break;
		}
		i++;
	}
	return result;
}

int ** enumNchooseK(int k, int n) {
	int **result = new int * [nChooseK(k, n)];
	for (int i = 0; i < nChooseK(k, n); i++) {
		result[i] = new int[k];
	}
	for (int j = 0; j < k; j++) {
		result[0][j] = j + 1;
	}
	for (int i = 1; i < nChooseK(k, n); i++) {
		int *nextEnum = next(result[i - 1], k, n);
		for (int j = 0; j < k; j++) {
			result[i][j] = nextEnum[j];
		}
	}
	return result;
}

int fact(int n) {
	if (n == 0) return 1;
	int result = 1;
	for (int i = 1; i <= n; i++) {
		result *= i;
	}
	return result;
}

int nChooseK(int k, int n) {
	if (k == 0) return 1;
	int result = 1;
	for (int i = 1; i < k + 1; i++) {
		result *= n - (k - i);
		result /= i;
	}
	return result;
}

bool isPerm(int a, int b) {
	std::string aAsStr = std::to_string(a); std::string bAsStr = std::to_string(b);
	if (aAsStr.length() != bAsStr.length()) return false;
	int *A = new int[10]; int *B = new int[10];
	for (int i = 0; i < aAsStr.length(); i++) {
		A[atoi(aAsStr.substr(i, 1).c_str())]++;
		B[atoi(bAsStr.substr(i, 1).c_str())]++;
	}
	for (int i = 0; i < 10; i++) {
		if (A[i] != B[i]) return false;
	}
	return true;
}

std::string * permutationsOfAString(std::string str) {
	if (str.length() == 2) {
		std::string *result = new std::string[2];
		result[0] = str;
		result[1] = str.substr(1) + str.substr(0, 1);
		return result;
	}
	return allNewStrings(permutationsOfAString(removeCharAt(str, str.length() - 1)), str.substr(str.length() - 1), fact(str.length() - 1)); 
}

void showStrings(std::string T[], int l) {
	for (int i = 0; i < l; i++) {
		std::cout << T[i] << std::endl;
	}
}